### PR TITLE
[REJECTED] Werror only on reporter.finish

### PR DESCRIPTION
**Pull request by som-snytt**
Give the tool runner more control over when warnings are
promoted to error. Instead of Werroring at the end of a run,
only Werror on finish.

### CODE REVIEW

**Comment by som-snytt** [comment]
@lrytz This is just a quickie that hooks into the new reporting stuff, which I don't fully understand yet.

Not sure this is wanted, but personally I like the way it looks: instead of adding an error to the count, it just says its bit at the end.

File: src/compiler/scala/tools/nsc/Reporting.scala [hunk 1]
```diff
@@ -109,16 +109,18 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
...
+        case Action.Error          => reporter.error(warning.pos, warning.msg)
+        case Action.Warning        => reporter.warning(warning.pos, warning.msg) ; werror()
+        case Action.WarningVerbose => reporter.warning(warning.pos, verbose) ; werror()
+        case Action.Info           => reporter.echo(warning.pos, warning.msg)
+        case Action.InfoVerbose    => reporter.echo(warning.pos, verbose)
         case a @ (Action.WarningSummary | Action.InfoSummary) =>
           val m = summaryMap(a, warning.category.summaryCategory)
-          if (!m.contains(warning.pos)) m.addOne((warning.pos, warning))
-        case Action.Silent =>
+          if (!m.contains(warning.pos)) { m.addOne((warning.pos, warning)) ; werror() }
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Reporting.scala]: should call `werror` only if `a == Action.WarningSummary`

**Review by lrytz — DISMISSED** [review]
I agree with this change in principle.

We should also consider how this interacts with zinc. For reasons I don't know, the zinc reporter intercepts warnings and [promotes them to errors under `-Werror`](https://github.com/sbt/zinc/blob/f608f13e171caa54a35d01aa97266fcba0d8bea7/internal/compiler-bridge/src/main/scala/xsbt/DelegatingReporter.scala#L180). This goes back to https://github.com/sbt/zinc/commit/7406a884b. Maybe when `Xwarnfatal` was introduced it was somehow not called when compiling through sbt?

Relatedly, are we sure that `reporter.finish()` is invoked when the compiler is driven by zinc?

Random observation: [`ReplReporterImpl.finish`](https://github.com/scala/scala/blob/8f20a7d84924939698717d26b3e0e5a5c0da829c/src/repl-frontend/scala/tools/nsc/interpreter/shell/Reporter.scala#L232-L235) should probably call `super.finish()`?

**Review by lrytz — COMMENTED** [review]
(hit "approve" by mistake, meant to "comment")

**Comment by som-snytt** [comment]
REPL behavior is currently bad:
```
✗ scala -Xlint -Werror
Welcome to Scala 2.13.1 (OpenJDK 64-Bit Server VM, Java 11.0.3).
Type in expressions for evaluation. Or try :help.

scala> @deprecated("noper","1") def f = 42
error: No warnings can be incurred under -Werror.
```
because it seems to be doing the wrong thing
```
package $line3 {
object $eval {
  lazy val $result = $line3.$read.$iw.$iw.f
  lazy val $print: _root_.java.lang.String =  {
    $line3.$read.$iw.$iw

""  + "f: Int\n"

  }
}}
```

File: src/compiler/scala/tools/nsc/Reporting.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala]: If I knew how it worked, I could write a test for it!

Comment (lrytz) [reply src/compiler/scala/tools/nsc/Reporting.scala]: From `-Wconf:help`

```
<action>
  - warning / w
  - warning-summary / ws (summary with the number of warnings, like for deprecations)
  - info / i             (infos are not counted as warnings and don't affect `-Werror`)
  - info-summary / is
```

too concise?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala]: I looked at the wconf tests but don't have even a little time for grokking; I'm not sure why warnings are issued in a different order now, I guess some sort of buffering.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/Reporting.scala]: Yes, that's due to `@nowarn`. The annotation is only recognized in typer, so any warning issued before is buffered until we can tell whether it's suppressed by an annotation.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala]: Right. When I wanted to defer something to after typer, the warning on widening, you said maybe just exploit this mechanism.

Comment (lrytz) [reply src/compiler/scala/tools/nsc/Reporting.scala]: Hmm, almost... There's `ContextReporter`, which is used as a proxy for reporting during type checking, see https://github.com/scala/scala/blob/8f20a7d84924939698717d26b3e0e5a5c0da829c/src/compiler/scala/tools/nsc/typechecker/Contexts.scala#L1584-L1595

Its main purpose, I think, is the ability to buffer messages in [`silent` mode](https://github.com/scala/scala/blob/8f20a7d84924939698717d26b3e0e5a5c0da829c/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L689)

Context reporters delegate to `currentRun.reporting`, aka `runReporting`. The reporting methods in `PerRunReporting` delegate in turn to the actual `Reporter`. In `PerRunReporting` we handle things like summarizing deprecations, and now also `-Wconf` and `@nowarn`.

The buffering of messages of pre-typer phases happens in `PerRunReporting`.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala]: Yes, sorry to make you explain, my attention is on something else.

**Comment by som-snytt** [comment]
IIRC `finish` is never called now.

**Pull request closed without merging**